### ISSUE TITLE
set(Type)::as_set(List, Set) doesn't take into account the `comparingp` protocol

### ISSUE THREAD

**Issue by aarroyoc** [LogtalkDotOrg/logtalk3#137]
After some checking with the info from the webinar, I found that `set(Type)::as_set/2` doesn't take into account custom comparisons. It simply does the same as set::as_set, which is call standard `sort/2`. 

Now, `set::as_set` is fine, but `set(Type)::as_set` should do custom comparisons like what sort does in this example: https://github.com/LogtalkDotOrg/logtalk3/blob/master/examples/sicstus/sorting.lgt even if it's slower.

I can try to contribute this fix

**Comment by pmoura** [comment]
As we discussed at the webinar, the type is currently only used for type-check the set elements. If we want to take consider user-defined types with their own ordering for the valid elements, all the other predicates, not just the `as_set/2` predicate will need to be defined for the `set(Type)` object.

**Comment by aarroyoc** [comment]
Yes, I know it's not a bug in the sense that something is not working as expected. The `set(Type)` object was designed this way. However, it doesn't feel ok if we use only the type information in one operation and not in others where this info can be useful. You're right that `set(Type)::as_set` wouldn't be the only predicate to change, `insert` and `insert_all` will surely need changes too.

**Comment by aarroyoc** [comment]
Another option would be creating a new implemention of `setp` which implements this behaviour

**Comment by pmoura** [comment]
I don't think that changing the predicates to use the sorting order provided by a custom type (instead of a standard term type) will introduce any breaking changes in practice. It will just make the predicates a bit slower, which I think is acceptable. So, let's got for this solution. Your contribution here is, of course, most welcome.
